### PR TITLE
Refactor GRPO loss computation for reuse

### PR DESCRIPTION
**Pull request by alexbowe**
# What does this PR do?

**Goal:** Refactor GRPO loss computation to support reuse.

**Context:** Custom trainers sometimes need tensors from GRPO loss. Today they must copy most of `_compute_loss`. This lets them reuse the loss calculation instead.

**Details:**

- `_compute_loss_state(...)` returns `_GRPOLossState`, containing the loss and intermediate tensors.
- `_log_loss_metrics(...)` keeps existing metric logging.
- `_compute_loss(...)` still returns only the scalar loss.

## Validation

- `uv run --extra test pytest -q 'tests/test_grpo_trainer.py::TestGRPOTrainer::test_train[trl-internal-testing/tiny-Qwen2ForCausalLM-2.5]' 'tests/test_grpo_trainer.py::TestGRPOTrainer::test_train_loss_types[bnpo-False]' 'tests/test_grpo_trainer.py::TestGRPOTrainer::test_train_loss_types[dr_grpo-False]' 'tests/test_grpo_trainer.py::TestGRPOTrainer::test_train_loss_types[dapo-False]' 'tests/test_grpo_trainer.py::TestGRPOTrainer::test_train_loss_types[cispo-False]' 'tests/test_grpo_trainer.py::TestGRPOTrainer::test_train_loss_types[sapo-False]' 'tests/test_grpo_trainer.py::TestGRPOTrainer::test_train_loss_types[luspo-False]' 'tests/test_grpo_trainer.py::TestGRPOTrainer::test_train_loss_types[vespo-False]'`
- `uvx ruff check trl/trainer/grpo_trainer.py`
- `uvx ruff format --check trl/trainer/grpo_trainer.py`

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that is the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [ ] Was this discussed or approved via a GitHub issue? Please add a link if so.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

No docs or new tests were added because this is an internal no-behavior-change refactor covered by existing GRPO trainer tests.

## AI writing disclosure

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

---

> [!NOTE]
> **Low Risk**
> Structural refactor in the training loss path with no intended training math changes; risk is mainly accidental metric or loss ordering regressions, mitigated by existing GRPO trainer tests.
> 
> **Overview**
> Splits **`GRPOTrainer`** loss handling so subclasses can reuse the full forward/loss math without duplicating **`_compute_loss`**.
> 
> Adds **`_GRPOLossState`** (scalar **`loss`** plus intermediates like **`per_token_logps`**, **`coef_1`**, **`per_token_kl`**, **`phi_seq`**, **`aux_loss`**). The former monolithic path becomes **`_compute_loss_state`** (compute + aggregate loss) and **`_log_loss_metrics`** (KL, entropy, clip ratios, VESPO/CISPO metrics, MoE aux). **`_compute_loss`** still returns only the scalar loss after logging.
> 
> Metric logging is slightly hardened with explicit **`None`** checks for optional KL, VESPO **`phi_seq`**, and aux loss before gather/append.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 5ea3e93680c07366d254ba14ab4fe1c69748442d. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by alexbowe** [comment]
Hey @AmineDiro, here's a small refactor to make GRPO loss computation reusable, for your consideration when you have time.

**Comment by AmineDiro** [comment]
Hi @alexbowe, thanks a lot for taking the time to put this together. The refactor itself is clean and the motivation is clear.

That said, I'm going to decline this one, and I'd rather explain the reasoning than just close it silently.

The fact that `_compute_loss` isn't reusable across trainers is a deliberate design choice, not an oversight. TRL trainers are intentionally self-contained: shared logic: generation, reward computation, the loss path, metric logging, weight syncing  is duplicated across trainers rather than factored into helpers or a base `class. _BaseTrainer` provides only
minimal utilities by design. The goal is that each trainer can be read top-to-bottom and modified in isolation, without a contributor having to trace shared abstractions or fear that a change ripples into other trainers.

For your use case (a custom trainer needing intermediate tensors), the intended path is to copy `_compute_loss` into your subclass and adapt it, duplication is accepted here; consistency is what we enforce.

Really appreciate the contribution and the thorough validation !
